cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll

```diff
@@ -1192,177 +1219,206 @@ private predicate nodeCand2(NodeExt node, boolean toReturn, boolean stored, Conf
-      preservesValue = false and
-      t = getErasedNodeTypeBound(node2)
-    ) and
-    node1 != node2 and
-    cc.relevantFor(node1.getEnclosingCallable()) and
-    not isUnreachableInCall(node1, cc.(LocalCallContextSpecificCall).getCall()) and
-    nodeCand(TNormalNode(node2), unbind(config))
-    or
-    exists(Node mid |
-      localFlowStepPlus(node1, mid, preservesValue, t, config, cc) and
-      localFlowStep(mid, node2, config) and
-      not mid instanceof CastNode and
-      nodeCand(TNormalNode(node2), unbind(config))
+      exists(Node mid |
+        localFlowStepPlus(node1, mid, preservesValue, t, config, cc) and
+        localFlowStep(mid, node2, config) and
+        not mid instanceof CastNode and
+        nodeCand2(TNormalNode(node2), unbind(config))
+      )
+      or
+      exists(Node mid |
+        localFlowStepPlus(node1, mid, _, _, config, cc) and
+        additionalLocalFlowStep(mid, node2, config) and
+        not mid instanceof CastNode and
+        preservesValue = false and
+        t = getErasedNodeTypeBound(node2) and
+        nodeCand2(TNormalNode(node2), unbind(config))
+      )
     )
+  }
+
+  /**
+   * Holds if `node1` can step to `node2` in one or more local steps and this
+   * path can occur as a maximal subsequence of local steps in a dataflow path.
+   */
+  pragma[nomagic]
+  predicate localFlowBigStep(
+    Node node1, Node node2, boolean preservesValue, DataFlowType t, Configuration config,
+    LocalCallContext callContext
+  ) {
+    localFlowStepPlus(node1, node2, preservesValue, t, config, callContext) and
+    localFlowExit(node2, config)
+  }
+
+  pragma[nomagic]
+  predicate localFlowBigStepExt(
+    NodeExt node1, NodeExt node2, boolean preservesValue, AccessPathFrontNil apf,
+    Configuration config
+  ) {
+    localFlowBigStep(node1.getNode(), node2.getNode(), preservesValue, apf.getType(), config, _)
     or
-    exists(Node mid |
-      localFlowStepPlus(node1, mid, _, _, config, cc) and
-      additionalLocalFlowStep(mid, node2, config) and
-      not mid instanceof CastNode and
-      preservesValue = false and
-      t = getErasedNodeTypeBound(node2) and
-      nodeCand(TNormalNode(node2), unbind(config))
-    )
-  )
+    additionalLocalFlowStepExt(node1, node2, apf.getType(), config) and
+    nodeCand2(node1, config) and
+    nodeCand2(node2, unbind(config)) and
+    preservesValue = false
+  }
 }
 
-/**
- * Holds if `node1` can step to `node2` in one or more local steps and this
- * path can occur as a maximal subsequence of local steps in a dataflow path.
- */
+private import LocalFlowBigStep
+
 pragma[nomagic]
-private predicate localFlowBigStep(
-  Node node1, Node node2, boolean preservesValue, DataFlowType t, Configuration config,
-  LocalCallContext callContext
-) {
-  localFlowStepPlus(node1, node2, preservesValue, t, config, callContext) and
-  localFlowExit(node2, config)
+private predicate readExtCand2_0(Content f, NodeExt node1, NodeExt node2, Configuration config) {
```

Comment:
Doing
```
pragma[nomagic]
private predicate readExtCand2(NodeExt node1, Content f, NodeExt node2, Configuration config) {
  readExt(node1, f, node2, config) and
  nodeCand2(node1, _, true, config) and
  nodeCand2(node2, unbind(config)) and
  readStoreCand(f, _, unbind(config))
}
```
does not work well:
```
[2020-03-18 13:02:12] (150s) Tuple counts for DataFlowImpl::readExtCand2#ffff:
                      11264   ~36%     {2} r1 = SCAN DataFlowImpl::readStoreCand#fff AS I OUTPUT I.<0>, I.<2>
                      1227910 ~57%     {5} r2 = JOIN r1 WITH DataFlowImpl::readExt#ffff_1023#join_rhs AS R ON FIRST 1 OUTPUT r1.<0>, r1.<1>, R.<1>, R.<2>, R.<3>
                      1227910 ~57%     {5} r3 = SELECT r2 ON r2.<1> >= r2.<4>
                      1227910 ~57%     {5} r4 = SELECT r3 ON r3.<1> <= r3.<4>
                      1227910 ~51%     {5} r5 = SCAN r4 OUTPUT r4.<2>, true, r4.<4>, r4.<0>, r4.<3>
                      478515  ~57%     {4} r6 = JOIN r5 WITH DataFlowImpl::nodeCand2#2#ffff_023#join_rhs AS R ON FIRST 3 OUTPUT r5.<4>, r5.<3>, r5.<0>, r5.<2>
                      433949  ~53%     {5} r7 = JOIN r6 WITH project#DataFlowImpl::nodeCand2#2#ffff AS R ON FIRST 1 OUTPUT r6.<1>, r6.<2>, r6.<0>, r6.<3>, R.<1>
                      433949  ~53%     {5} r8 = SELECT r7 ON r7.<4> >= r7.<3>
                      433949  ~53%     {5} r9 = SELECT r8 ON r8.<4> <= r8.<3>
                      433949  ~54%     {4} r10 = SCAN r9 OUTPUT r9.<1>, r9.<0>, r9.<2>, r9.<3>
                                       return r10
```

Reply:
Hmm, is that duplication from the `_` in `readStoreCand`?  Actually, it looks like the boolean column in `readStoreCand` isn't even used, so that could be turned into an `exists` inside `readStoreCand`.

Reply:
Removing the Boolean from `readStoreCand` gives a better pipeline, though the join-order is not quite as good as the folded version (but I can live with that):
```
[2020-03-18 18:02:54] (117s) Tuple counts for DataFlowImpl::readExtCand2#ffff:
                      783905  ~0%     {5} r1 = JOIN DataFlowImpl::readStoreCand#ff AS L WITH DataFlowImpl::readExt#ffff_1023#join_rhs AS R ON FIRST 1 OUTPUT L.<0>, L.<1>, R.<1>, R.<2>, R.<3>
                      783905  ~0%     {5} r2 = SELECT r1 ON r1.<1> >= r1.<4>
                      783905  ~0%     {5} r3 = SELECT r2 ON r2.<1> <= r2.<4>
                      783905  ~0%     {5} r4 = SCAN r3 OUTPUT r3.<2>, true, r3.<4>, r3.<0>, r3.<3>
                      303496  ~0%     {4} r5 = JOIN r4 WITH DataFlowImpl::nodeCand2#2#ffff_023#join_rhs AS R ON FIRST 3 OUTPUT r4.<4>, r4.<3>, r4.<0>, r4.<2>
                      274055  ~0%     {5} r6 = JOIN r5 WITH project#DataFlowImpl::nodeCand2#2#ffff AS R ON FIRST 1 OUTPUT r5.<1>, r5.<2>, r5.<0>, r5.<3>, R.<1>
                      274055  ~0%     {5} r7 = SELECT r6 ON r6.<4> >= r6.<3>
                      274055  ~0%     {5} r8 = SELECT r7 ON r7.<4> <= r7.<3>
                      274055  ~0%     {4} r9 = SCAN r8 OUTPUT r8.<1>, r8.<0>, r8.<2>, r8.<3>
                                      return r9
```